homeassistant/components/recorder/statistics.py

```diff
@@ -0,0 +1,64 @@
+"""Statistics helper."""
+from __future__ import annotations
+
+from datetime import date, datetime, timedelta
+import logging
+from typing import TYPE_CHECKING
+
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN
+from .models import Statistics
+from .util import retriable_database_job, session_scope
+
+if TYPE_CHECKING:
+    from . import Recorder
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_start_time(period: str) -> datetime.datetime:
+    """Return start time give a period."""
+    if period == "daily":
+        # Get start and end times at local midnigt. This will result in a period of 23 hours
+        # at start of DST and a period of 25 hours at end of DST
+        yesterday = date.today() - timedelta(days=1)
+        start = datetime.combine(yesterday, datetime.min.time()).astimezone(
+            dt_util.NATIVE_UTC
+        )
+    else:
+        last_hour = dt_util.utcnow() - timedelta(hours=1)
+        start = last_hour.replace(minute=0, second=0, microsecond=0)
+    return start
+
+
+@retriable_database_job
+def compile_statistics(
+    instance: Recorder, period: str, start: datetime.datetime
+) -> bool:
+    """Compile statistics."""
+    start = dt_util.as_utc(start)
+    if period == "daily":
+        end = start + timedelta(days=1)
```

Comment:
It feels a bit weird that we have a `get_start_time()` result passed into this method, and then calculate `end` here. I do understand the reasoning, as we can now guarantee the length of a period based on the period. It's just, well something I noticed :D Probably ok as-is. Alternative is to create a `duration` helper object that takes period and start.